### PR TITLE
Update llvm to r302975

### PR DESCRIPTION
**Pull request by vgvassilev**

### CODE REVIEW

**Comment by vgvassilev** [comment]
@phsft-bot build with flags -Dvc=OFF -Dimt=ON -Dccache=OFF

**Comment by vgvassilev** [comment]
@phsft-bot build just on centos7/gcc49, ubuntu14/native with flags -Dccache=OFF

**Comment by vgvassilev** [comment]
@phsft-bot build with flags -Dccache=OFF

**Comment by martinmine** [comment]
@phsft-bot build with flags -Dccache=OFF

**Comment by martinmine** [comment]
@phsft-bot build with flags -Dccache=OFF

**Comment by martinmine** [comment]
@phsft-bot build with flags -Dccache=OFF

**Comment by vgvassilev** [comment]
@phsft-bot build!

**Comment by marsupial** [comment]
**778b83c Create a virtual file entry for each input line.**

Will this mess up clang's file caching so all filenames following the pattern **input_file_N** won't be include-able?

**Comment by vgvassilev** [comment]
@marsupial, I can make the inclusion of `ConvertUTF.h` conditionally included if `LLVM_UTF8` is set. It is not clear to me what would be the benefit, though.

**Comment by Teemperor** [comment]
@phsft-bot build just on mac1012/native

**Comment by Teemperor** [comment]
@phsft-bot build just on mac1012/native

**Comment by marsupial** [comment]
> @marsupial, I can make the inclusion of ConvertUTF.h conditionally included if LLVM_UTF8 is set. It is not clear to me what would be the benefit, though.

It already is that way. Look a few lines below the include to see it.  The benefit is to use the C++ lib to do it when it can so client conversion will match as well as not allocate 4x the string length which is what the LLVM routines do.

> Can you give an example of which case you are thinking go?

Have a memory of  virtual file names being inserted into the cache so clang thinks it already hold a buffer to them, which would make the following not work:
```
$ int Val = 5; // input_line_1
$ #include  "input_line_1" // input_line_2
$ #include  "input_line_3" // input_line_3
$ #include  "input_line_5" // input_line_4
$ Val = 6; // input_line_5
```

**Comment by vgvassilev** [comment]
I overlooked the issue with the llvm conversion routines. I've inserted a better fix.

For the second issue, I think it might be a good idea to do it as a post PR merge commit because it requires some changes in the test suites of both ROOT and cling.

**Comment by martinmine** [comment]
@phsft-bot build just on slc6/icc17

File: core/clingutils/src/TClingUtils.cxx [hunk 1]
```diff
@@ -41,6 +41,7 @@
 #include "clang/Lex/HeaderSearch.h"
 #include "clang/Lex/ModuleMap.h"
 #include "clang/Lex/Preprocessor.h"
+#include "clang/Lex/PreprocessorOptions.h"
```
Comment (Axel-Naumann) [inline core/clingutils/src/TClingUtils.cxx]: What is that needed for?

File: interpreter/cling/include/cling/Interpreter/Interpreter.h [hunk 2]
```diff
@@ -12,8 +12,6 @@
 
 #include "cling/Interpreter/InvocationOptions.h"
 
-#include "llvm/ADT/StringRef.h"
```
Comment (Axel-Naumann) [inline interpreter/cling/include/cling/Interpreter/Interpreter.h]: Great! But shouldn't this also mean that we can reduce the number of headers that gets copied by the build  system?

File: core/clingutils/src/TClingUtils.cxx (hunk 1 above)
Comment (vgvassilev) [inline core/clingutils/src/TClingUtils.cxx]: Didn't really see. It fixed compilation.

File: interpreter/cling/include/cling/Interpreter/Interpreter.h (hunk 2 above)
Comment (Axel-Naumann) [inline interpreter/cling/include/cling/Interpreter/Interpreter.h]: It's likely missing many removals. Given that `StringRef.h` was the only llvm header we `#include`d in `Interpreter.h` I'd expect that we can remove *all* llvm headers from that list! I.e. empty it. I.e. remove that whole feature.

Comment (vgvassilev) [inline interpreter/cling/include/cling/Interpreter/Interpreter.h]: I see. I think this is a future improvement somewhat outside of the scope of this PR.

**Comment by vgvassilev** [comment]
@phsft-bot build also on slc6/icc17

**Comment by vgvassilev** [comment]
@phsft-bot build also on slc6/icc17

**Comment by vgvassilev** [comment]
@phsft-bot build also on slc6/icc17

**Comment by vgvassilev** [comment]
The update of the LLVM exposed an issue #614 which is outside of the scope of this PR.

File: interpreter/cling/include/cling/Interpreter/Interpreter.h (hunk 2 above)
Comment (Axel-Naumann) [inline interpreter/cling/include/cling/Interpreter/Interpreter.h]: That's fine, please submit a subsequent PR then.

**Comment by marsupial** [comment]
Could you fix **cpt.py** in the cling repo which is pulling from "https://raw.githubusercontent.com/vgvassilev/cling/master/LastKnownGoodLLVMSVNRevision.txt" and breaking Travis.

**Comment by vgvassilev** [comment]
Done. Thanks!

**Comment by marsupial** [comment]
And Travis should remove gcc-4.7 builds as LLVM now requires 4.8.

**Comment by marsupial** [comment]
@vgvassilev Re: **778b83c Create a virtual file entry for each input line.**
Here's a simple failure:
```
./bin/cling -Xclang -verify
[$]for (int i=0;i<1;++i) { val; } // expected-error {{use of undeclared identifier 'val'}}
[$].q
fatal error: cannot open file 'input_line_3': No such file or directory
```

**Comment by marsupial** [comment]
@vgvassilev https://github.com/root-project/cling/pull/146/commits/208cf07c40c0113904a250726bac3ef82efa5c16 fixes the problem

**Comment by pcanal** [comment]
The new llvm version 'seems' to bring 'new' valgrind warnings.  For example:
```valgrind --suppresions=$ROOTSYS/etc/valgrind-root.supp root.exe -b -l -q```
outputs
```
==9328== Conditional jump or move depends on uninitialised value(s)
==9328==    at 0xAD44B75: llvm::sys::path::filename(llvm::StringRef, llvm::sys::path::Style) (in /local2/scratch/pcanal/root_builds/master/c++11/lib/libCling.so)
==9328==    by 0x91D1526: clang::CodeGen::CodeGenModule::EmitCXXGlobalInitFunc() (in /local2/scratch/pcanal/root_builds/master/c++11/lib/libCling.so)
==9328==    by 0x90F366F: clang::CodeGen::CodeGenModule::Release() (in /local2/scratch/pcanal/root_builds/master/c++11/lib/libCling.so)
==9328==    by 0x903F906: clang::CodeGeneratorImpl::HandleTranslationUnit(clang::ASTContext&) (in /local2/scratch/pcanal/root_builds/master/c++11/lib/libCling.so)
==9328==    by 0x8B9E39F: cling::IncrementalParser::codeGenTransaction(cling::Transaction*) (IncrementalParser.cpp:544)
==9328==    by 0x8B9E078: cling::IncrementalParser::commitTransaction(llvm::PointerIntPair<cling::Transaction*, 2u, cling::IncrementalParser::EParseResult, llvm::PointerLikeTypeTraits<cling::Transaction*>, llvm::PointerIntPairInfo<cling::Transaction*, 2u
, llvm::PointerLikeTypeTraits<cling::Transaction*> > >&, bool) (IncrementalParser.cpp:481)
==9328==    by 0x8B255F4: cling::Interpreter::Interpreter(int, char const* const*, char const*, bool, cling::Interpreter const*) (Interpreter.cpp:245)
==9328==    by 0x894B3B8: cling::Interpreter::Interpreter(int, char const* const*, char const*, bool) (Interpreter.h:308)
==9328==    by 0x8A3B5B1: TCling::TCling(char const*, char const*) (TCling.cxx:1116)
==9328==    by 0x8A38697: CreateInterpreter (TCling.cxx:593)
==9328==    by 0x517F717: TROOT::InitInterpreter() (TROOT.cxx:1974)
==9328==    by 0x517983F: ROOT::Internal::GetROOT2() (TROOT.cxx:364)

==9328== Use of uninitialised value of size 8
==9328==    at 0xACFA1A0: llvm::StringRef::find_last_of(llvm::StringRef, unsigned long) const (in /local2/scratch/pcanal/root_builds/master/c++11/lib/libCling.so)
==9328==    by 0xAD41F86: (anonymous namespace)::filename_pos(llvm::StringRef, llvm::sys::path::Style) (in /local2/scratch/pcanal/root_builds/master/c++11/lib/libCling.so)
==9328==    by 0xAD44A83: llvm::sys::path::filename(llvm::StringRef, llvm::sys::path::Style) (in /local2/scratch/pcanal/root_builds/master/c++11/lib/libCling.so)
==9328==    by 0x91D1526: clang::CodeGen::CodeGenModule::EmitCXXGlobalInitFunc() (in /local2/scratch/pcanal/root_builds/master/c++11/lib/libCling.so)
==9328==    by 0x90F366F: clang::CodeGen::CodeGenModule::Release() (in /local2/scratch/pcanal/root_builds/master/c++11/lib/libCling.so)
==9328==    by 0x903F906: clang::CodeGeneratorImpl::HandleTranslationUnit(clang::ASTContext&) (in /local2/scratch/pcanal/root_builds/master/c++11/lib/libCling.so)
==9328==    by 0x8B9E39F: cling::IncrementalParser::codeGenTransaction(cling::Transaction*) (IncrementalParser.cpp:544)
==9328==    by 0x8B9E078: cling::IncrementalParser::commitTransaction(llvm::PointerIntPair<cling::Transaction*, 2u, cling::IncrementalParser::EParseResult, llvm::PointerLikeTypeTraits<cling::Transaction*>, llvm::PointerIntPairInfo<cling::Transaction*, 2u, llvm::PointerLikeTypeTraits<cling::Transaction*> > >&, bool) (IncrementalParser.cpp:481)
==9328==    by 0x8B255F4: cling::Interpreter::Interpreter(int, char const* const*, char const*, bool, cling::Interpreter const*) (Interpreter.cpp:245)
==9328==    by 0x894B3B8: cling::Interpreter::Interpreter(int, char const* const*, char const*, bool) (Interpreter.h:308)
==9328==    by 0x8A3B5B1: TCling::TCling(char const*, char const*) (TCling.cxx:1116)
==9328==    by 0x8A38697: CreateInterpreter (TCling.cxx:593)

==9328== 
==9328== Use of uninitialised value of size 8
==9328==    at 0x91D15BB: clang::CodeGen::CodeGenModule::EmitCXXGlobalInitFunc() (in /local2/scratch/pcanal/root_builds/master/c++11/lib/libCling.so)
==9328==    by 0x90F366F: clang::CodeGen::CodeGenModule::Release() (in /local2/scratch/pcanal/root_builds/master/c++11/lib/libCling.so)
==9328==    by 0x903F906: clang::CodeGeneratorImpl::HandleTranslationUnit(clang::ASTContext&) (in /local2/scratch/pcanal/root_builds/master/c++11/lib/libCling.so)
==9328==    by 0x8B9E39F: cling::IncrementalParser::codeGenTransaction(cling::Transaction*) (IncrementalParser.cpp:544)
==9328==    by 0x8B9E078: cling::IncrementalParser::commitTransaction(llvm::PointerIntPair<cling::Transaction*, 2u, cling::IncrementalParser::EParseResult, llvm::PointerLikeTypeTraits<cling::Transaction*>, llvm::PointerIntPairInfo<cling::Transaction*, 2u, llvm::PointerLikeTypeTraits<cling::Transaction*> > >&, bool) (IncrementalParser.cpp:481)
==9328==    by 0x8B255F4: cling::Interpreter::Interpreter(int, char const* const*, char const*, bool, cling::Interpreter const*) (Interpreter.cpp:245)
==9328==    by 0x894B3B8: cling::Interpreter::Interpreter(int, char const* const*, char const*, bool) (Interpreter.h:308)
==9328==    by 0x8A3B5B1: TCling::TCling(char const*, char const*) (TCling.cxx:1116)
==9328==    by 0x8A38697: CreateInterpreter (TCling.cxx:593)
==9328==    by 0x517F717: TROOT::InitInterpreter() (TROOT.cxx:1974)
==9328==    by 0x517983F: ROOT::Internal::GetROOT2() (TROOT.cxx:364)
==9328==    by 0x5179A3F: ROOT::GetROOT() (TROOT.cxx:488)
==9328== 
```

**Comment by pcanal** [comment]
Another longer set (see attachement array.vallog.txt
) from


```
cd roottest/root/tree/array
valgrind --suppression=$ROOTSYS/etc/vagrind-root.supp --log-file=array.vallog root.exe -e "#define ClingWorkAroundPrintfIssues" -e "#define ClingWorkAroundLackOfModule" -e "#define ClingWorkAroundProxyConfusion" -e "#define ClingWorkAroundMissingDynamicScope" -e "#define ClingWorkAroundMultipleInclude" -e "#define ClingWorkAroundAutoParseUsingNamespace" -e "#define ClingWorkAroundExtraParensWithImplicitAuto" -e "#define ClingWorkAroundNoPrivateClassIO" -e "#define ClingWorkAroundBrokenUnnamedReturn" -e "#define ClingWorkAroundUnnamedDetection2" -e "#define ClingWorkAroundUnnamedInclude" -e "#define ClingWorkAroundJITfullSymbolResolution" -e "#define ClingWorkAroundNoDotNamespace" -e "#define ClingWorkAroundNoDotInclude" -e "#define ClingWorkAroundUnnamedIncorrectFileLoc" -e "#define ClingWorkAroundTClassUpdateDouble32" -e "#define ClingWorkAroundUnloadingVTABLES" -e 'gROOT->ProcessLine(".L TestObj.cpp+");' -q -l -b runobjArrayTest.C
```

**Comment by pcanal** [comment]
humm .. upload fails so far ... (Send it to vassil by mail)

**Comment by pcanal** [comment]
[array.vallog.txt.gz](https://github.com/root-project/root/files/1063419/array.vallog.txt.gz)

**Comment by pcanal** [comment]
now succedded.

**Comment by Axel-Naumann** [comment]
I have fixed the valgrind error from `clang::CodeGen::CodeGenModule::EmitCXXGlobalInitFunc()`

**Pull request merged**